lib/ukstore/include/uk/store.h

```diff
@@ -0,0 +1,452 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __UK_STORE_H__
+#define __UK_STORE_H__
+
+#include <uk/config.h>
+#include <uk/essentials.h>
+#include <string.h>
+#include <uk/list.h>
+#include <uk/arch/atomic.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if !CONFIG_LIBUKSTORE
+#define uk_store_libid(libname) (-1)
+
+#define UK_STORE_INITREG_FOLDER(fldr)
+
+#define UK_STORE_INITREG_ENTRY(entry, e_type, e_get, e_set)
+
+#define uk_store_entry_init(entry, e_name, e_type, getter, setter)
+
+#define uk_store_folder_entry_init(folder_entry, new_entry)
+
+#define uk_store_folder_init(folder, f_name)
+
+#define uk_store_add_folder(folder)
+
+#define uk_store_del_folder(folder)
+
+#define uk_store_add_folder_entry(folder, folder_entry)
+
+#define uk_store_del_folder_entry(folder_entry)
+
+#else /* !CONFIG_LIBUKSTORE */
+
+/* Library-specific information generated at compile-time */
+#include <uk/_store_libs.h>
+
+/* Do not call directly */
+#define _uk_store_libid(libname)	\
+	__UK_STORE_ ## libname
+
+/**
+ * Returns the id of a given library.
+ *
+ * @return the id
+ */
+#define uk_store_libid(libname)	\
+	_uk_store_libid(libname)
+
+#endif /* CONFIG_LIBUKSTORE */
+
+/**
+ * Returns the name of the library
+ *
+ * @return this libraries' name
+ */
+#define uk_store_self_name()	\
+	__LIBNAME__
+
+/**
+ * Returns the id of the place it was called from.
+ *
+ * @return this libraries' id
+ */
+#define uk_store_libid_self()	\
+	uk_store_libid(uk_store_self_name())
+
+/* All basic types that exist - use these when interacting with the API */
+enum uk_store_entry_basic_type {
+	none,
+	s8,
+	u8,
+	s16,
+	u16,
+	s32,
+	u32,
+	s64,
+	u64,
+	uptr
```

Comment:
I tested it and there aren't any clashes. The ones in `uklibparams` have this form: `_LIB_PARAM___s32`. Still I am up for this change if you think it would help in the long run.

Reply:
I think you should be prepared for the worst case. The reason is that your library will likely be included by many other libraries that we start integrating. They may have own or (more likely) no naming conventions. This means that a clash becomes much more likely. I am fine in calling them for instance `_LIB_UKSTORE__s32` so that they are inline with `uklibparams` but we should namespace.